### ISSUE TITLE
Copy/Paste of the PromiseLike interface incompatible with itself

### ISSUE THREAD

**Issue by paul-marechal** [microsoft/TypeScript#31264]
**TypeScript Version:**  3.4.5

**Search Terms:** `promiselike`

**Code**

Reference from `lib.es5.d.ts`:

```ts
interface PromiseLike<T> {
    /**
     * Attaches callbacks for the resolution and/or rejection of the Promise.
     * @param onfulfilled The callback to execute when the Promise is resolved.
     * @param onrejected The callback to execute when the Promise is rejected.
     * @returns A Promise for the completion of which ever callback is executed.
     */
    then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): PromiseLike<TResult1 | TResult2>;
}
```

My `eXtendedPromiseLike` interface (actually a copy/paste for testing purposes):

```ts
export interface XPromiseLike<T> {

    then<TResult1 = T, TResult2 = never>(
        onfulfilled?: ((value: T) => TResult1 | XPromiseLike<TResult1>) | undefined | null,
        onrejected?: ((reason: any) => TResult2 | XPromiseLike<TResult2>) | undefined | null
    ): XPromiseLike<TResult1 | TResult2>;

}
```

**Expected behavior:**

Somehow for the types to be compatible?

**Actual behavior:**

```
Type 'XPromiseLike<number>' is not assignable to type 'PromiseLike<number>'.
  Types of property 'then' are incompatible.
    Type '<TResult1 = number, TResult2 = never>(onfulfilled?: (value: number) => TResult1 | XPromiseLike<TResult1>, onrejected?: (reason: any) => TResult2 | XPromiseLike<TResult2>) => XPromiseLike<TResult1 | TResult2>' is not assignable to type '<TResult1 = number, TResult2 = never>(onfulfilled?: (value: number) => TResult1 | PromiseLike<TResult1>, onrejected?: (reason: any) => TResult2 | PromiseLike<TResult2>) => PromiseLike<TResult1 | TResult2>'.
      Types of parameters 'onrejected' and 'onrejected' are incompatible.
        Type 'TResult2 | PromiseLike<TResult2>' is not assignable to type 'TResult2 | XPromiseLike<TResult2>'.
          Type 'PromiseLike<TResult2>' is not assignable to type 'TResult2 | XPromiseLike<TResult2>'.
            Type 'PromiseLike<TResult2>' is not assignable to type 'XPromiseLike<TResult2>'.
              Types of property 'then' are incompatible.
                Type '<TResult1 = TResult2, TResult2 = never>(onfulfilled?: (value: TResult2) => TResult1 | PromiseLike<TResult1>, onrejected?: (reason: any) => TResult2 | PromiseLike<TResult2>) => PromiseLike<TResult1 | TResult2>' is not assignable to type '<TResult1 = TResult2, TResult2 = never>(onfulfilled?: (value: TResult2) => TResult1 | XPromiseLike<TResult1>, onrejected?: (reason: any) => TResult2 | XPromiseLike<TResult2>) => XPromiseLike<TResult1 | TResult2>'.
                  Types of parameters 'onrejected' and 'onrejected' are incompatible.
                    Type 'TResult2 | XPromiseLike<TResult2>' is not assignable to type 'TResult2 | PromiseLike<TResult2>'.
                      Type 'XPromiseLike<TResult2>' is not assignable to type 'TResult2 | PromiseLike<TResult2>'.
                        Type 'XPromiseLike<TResult2>' is not assignable to type 'PromiseLike<TResult2>'.
                          Types of property 'then' are incompatible.
                            Type '<TResult1 = TResult2, TResult2 = never>(onfulfilled?: (value: TResult2) => TResult1 | XPromiseLike<TResult1>, onrejected?: (reason: any) => TResult2 | XPromiseLike<TResult2>) => XPromiseLike<TResult1 | TResult2>' is not assignable to type '<TResult1 = TResult2, TResult2 = never>(onfulfilled?: (value: TResult2) => TResult1 | PromiseLike<TResult1>, onrejected?: (reason: any) => TResult2 | PromiseLike<TResult2>) => PromiseLike<TResult1 | TResult2>'.
                              Types of parameters 'onrejected' and 'onrejected' are incompatible.
                                Type 'TResult2 | PromiseLike<TResult2>' is not assignable to type 'TResult2 | XPromiseLike<TResult2>'.
                                  Type 'PromiseLike<TResult2>' is not assignable to type 'TResult2 | XPromiseLike<TResult2>'.
```

**Playground Link:** [`PromiseLike<T>` vs `XPromiseLike<T>`](https://www.typescriptlang.org/play/#src=interface%20XPromiseLike%3CT%3E%20%7B%0D%0A%20%20%20%20%2F**%0D%0A%20%20%20%20%20*%20Attaches%20callbacks%20for%20the%20resolution%20and%2For%20rejection%20of%20the%20Promise.%0D%0A%20%20%20%20%20*%20%40param%20onfulfilled%20The%20callback%20to%20execute%20when%20the%20Promise%20is%20resolved.%0D%0A%20%20%20%20%20*%20%40param%20onrejected%20The%20callback%20to%20execute%20when%20the%20Promise%20is%20rejected.%0D%0A%20%20%20%20%20*%20%40returns%20A%20Promise%20for%20the%20completion%20of%20which%20ever%20callback%20is%20executed.%0D%0A%20%20%20%20%20*%2F%0D%0A%20%20then%3CTResult1%20%3D%20T%2C%20TResult2%20%3D%20never%3E(%0D%0A%20%20%20%20onfulfilled%3F%3A%20((value%3A%20T)%20%3D%3E%20TResult1%20%7C%20XPromiseLike%3CTResult1%3E)%20%7C%20undefined%20%7C%20null%2C%0D%0A%20%20%20%20onrejected%3F%3A%20((reason%3A%20any)%20%3D%3E%20TResult2%20%7C%20XPromiseLike%3CTResult2%3E)%20%7C%20undefined%20%7C%20null%0D%0A%20%20)%3A%20XPromiseLike%3CTResult1%20%7C%20TResult2%3E%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20const%20a%3A%20XPromiseLike%3Cnumber%3E%0D%0Aconst%20b%3A%20PromiseLike%3Cnumber%3E%20%3D%20a%0D%0A)

**Related Issues:** -

---

Originally I was trying to add a template parameter to specify the type of the error passed to the `onrejected` callback, but then it spiraled down into this issue. Maybe I am doing something wrong, but I don't understand why a copy/paste is failing.

**Comment by fatcerberus** [comment]
Okay this is a really weird bug.  Copy-and-pasting `PromiseLike` into the playground as-is works fine.  Simply renaming it (right click -> change all occurrences) breaks it, which makes no sense under structural typing.

**Comment by fatcerberus** [comment]
I don't know about the recursion, but the part I find weird is that you can paste `PromiseLike` as-is into the playground and it works fine, but renaming it with literally no other changes breaks it.  I didn't even think the compiler *cared* about the names of types!

**Comment by marechal-p** [comment]
I assume that if the compiler can understand that two types (based on what you wrote, so the names) actually point to the same structure, then it might be able to take shortcuts and maybe just compare the template types. But here since we have two different interfaces, it must try to check the structure to evaluate if they are compatible, and I think this is where it loses it due to some recursion bug?

**Comment by weswigham** [comment]
Related, maybe: #31570

**Comment by weswigham** [comment]
This is fixed as of TS 4.9~
